### PR TITLE
Porting DataMemberFieldConverter into core

### LINKED ISSUE #1542: WinForms Interop: XamlParseException on .NET Core
@thomasclaudiushuber commented on [Wed Jul 24 2019](https://github.com/dotnet/wpf/issues/1366)

* .NET Core Version: 3.0.100-preview8-013410
* Windows version: 1903
* Does the bug reproduce also in WPF for .NET Framework 4.8?: No
 
 **Problem description:**
Creating a WinForms DataGridView in WPF XAML throws XamlParseException on .NET Core. This happens for the DataPropertyName property of the DataGridViewTextBoxColumn:

```
   <winForms:DataGridViewTextBoxColumn HeaderText="Firstname"
                                                        DataPropertyName="FirstName"/>
```
 
 **Actual behavior:** 

```
System.Windows.Markup.XamlParseException
  HResult=0x80131501
  Message=Invalid metadata for attribute 'System.ComponentModel.TypeConverterAttribute' on 'System.String DataPropertyName'. Expected '1' argument(s) of type 'System.Type'.
  Source=PresentationFramework
  StackTrace:
   at System.Windows.Markup.XamlReader.RewrapException(Exception e, IXamlLineInfo lineInfo, Uri baseUri)
   at System.Windows.Markup.WpfXamlLoader.Load(XamlReader xamlReader, IXamlObjectWriterFactory writerFactory, Boolean skipJournaledProperties, Object rootObject, XamlObjectWriterSettings settings, Uri baseUri)
   at System.Windows.Markup.WpfXamlLoader.LoadBaml(XamlReader xamlReader, Boolean skipJournaledProperties, Object rootObject, XamlAccessLevel accessLevel, Uri baseUri)
   at System.Windows.Markup.XamlReader.LoadBaml(Stream stream, ParserContext parserContext, Object parent, Boolean closeStream)
   at System.Windows.Application.LoadComponent(Object component, Uri resourceLocator)
   at WinFormsInWpf.MainWindow.InitializeComponent() in C:\Users\thoma\source\repos\WinFormsInWpf.Framework\WinFormsInWpf.Core\MainWindow.xaml:line 1
   at WinFormsInWpf.MainWindow..ctor() in C:\Users\thoma\source\repos\WinFormsInWpf.Framework\WinFormsInWpf.Core\MainWindow.xaml.cs:line 9

Inner Exception 1:
XamlSchemaException: Invalid metadata for attribute 'System.ComponentModel.TypeConverterAttribute' on 'System.String DataPropertyName'. Expected '1' argument(s) of type 'System.Type'.
```

  **Expected behavior:**
 It should just work like in .NET Framework

 **Minimal repro:**
I've created a minimal repro solution here that contains a working project that targets .NET Framework and a failing project that targets .NET Core: https://github.com/thomasclaudiushuber/WpfWinFormsXamlParseIssue

 

---

@rladuca commented on [Mon Jul 29 2019](https://github.com/dotnet/wpf/issues/1366#issuecomment-516208100)

This looks like possibly a WinForms bug.  From what I am seeing, we're choking on trying to use `DataMemberFieldConverter` which is marked on a [few WinForms types](https://github.com/dotnet/winforms/search?q=DataMemberFieldConverter&unscoped_q=DataMemberFieldConverter).

This used to live in System.Design.  When I look at System.Design in WindowsDesktop, I see it's a facade assembly now, but there is no type forward for `DataMemberFieldConverter`.  This causes a Type.GetType call to fail [here](https://github.com/dotnet/wpf/blob/2cbb1ad9759c32dc575c7537057a29ee7da2e1b2/src/Microsoft.DotNet.Wpf/src/System.Xaml/System/Xaml/Schema/XamlNamespace.cs#L188).

@zsd4yr @merriemcgaw @Tanya-Solyanik @AdamYoblick Does this ring any bells or am I off base?

---

@merriemcgaw commented on [Mon Jul 29 2019](https://github.com/dotnet/wpf/issues/1366#issuecomment-516208378)

@dreddy-work can you take a look? You did a lot of the System.Design stuff on our end. Seems possible to me.

---

@dreddy-work commented on [Mon Jul 29 2019](https://github.com/dotnet/wpf/issues/1366#issuecomment-516219875)

DataMemberFieldConverter is not yet ported to core. I will make those changes and get the PR out.

---

@dreddy-work commented on [Tue Jul 30 2019](https://github.com/dotnet/wpf/issues/1366#issuecomment-516416035)

@rladuca , DataMemberFieldConverter was internal class and we don't see typeforward even after we port it. How exactly this typeconverter is used in wpf? 

---

@weltkante commented on [Tue Jul 30 2019](https://github.com/dotnet/wpf/issues/1366#issuecomment-516427394)

@dreddy-work Converters don't need to be public in order to be referenced from attributes, in WinForms you are exporting TypeConverter attributes referencing System.Design by concatenating strings (see the [link](https://github.com/dotnet/winforms/search?q=DataMemberFieldConverter&unscoped_q=DataMemberFieldConverter) from @rladuca above).

WPF finds the attribute and wants to resolve the string, but there is not type forwarding in place.

---

@dreddy-work commented on [Tue Jul 30 2019](https://github.com/dotnet/wpf/issues/1366#issuecomment-516429814)

Thanks. Will port this and make sure right assembly is referenced. @merriemcgaw  can you help move this issue to winforms repo?

---

@rladuca commented on [Tue Jul 30 2019](https://github.com/dotnet/wpf/issues/1366#issuecomment-516507724)

@dreddy-work It's probably a good idea to sweep for any TypeConverters you are using, WPF will break on using any WinForm control where there is a similar state to this bug.

### PR DESCRIPTION
**Pull request by dreddy-work**
Porting DataMemberFieldConverter to System.Windows.Forms.Design.dll and changing the references from System.Design to this assembly wherever required. 

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/1545)

Fixes #1542
**Proposed changes**
- Porting DataMemberFieldConverter into core. 

**Customer Impact**
 - Some of the winforms properties have added this converter as member attribute and thus use of those properties, by third party applications, resulting in 'MissingMethodExceptions'.

**Regression?**
Yes

**Risk**
None

**Screenshots 
Before**

image

**After**

No Exception.

### CODE REVIEW

**Comment by dreddy-work** [comment]
tests need to be cleaned up.

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewColumn.cs [hunk 1]
```diff
@@ -186,7 +186,7 @@ public override ContextMenuStrip ContextMenuStrip
         [
             Browsable(true),
             DefaultValue(""),
-            TypeConverter("System.Windows.Forms.Design.DataMemberFieldConverter, " + AssemblyRef.SystemDesign),
+            TypeConverter("System.Windows.Forms.Design.DataMemberFieldConverter, " + AssemblyRef.SystemWinformsDesign),
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewColumn.cs]: I don't believe this is a correct way of doing this.
We need to forward the type to System.Design similar to how it was done in #1299 (https://github.com/dotnet/winforms/blob/master/src/System.Design/src/System.Design.Forwards.cs)

File: src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/DataMemberFieldConverter.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+        public override object ConvertFrom(ITypeDescriptorContext context, CultureInfo culture, object value)
+        {
+            return value != null && value.Equals(SR.None) ? string.Empty : value;
+        }
+
+        public override object ConvertTo(ITypeDescriptorContext context, CultureInfo culture, object value, Type destinationType)
+        {
+            if (destinationType == typeof(string) && (value == null || value.Equals(string.Empty)))
+                return SR.None_lc;
```
Comment (RussKie) [inline src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/DataMemberFieldConverter.cs]: missing `{ }`

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewColumn.cs (hunk 1 above)
Comment (dreddy-work) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewColumn.cs]: I was looking at that. What was the reason for that?

Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewColumn.cs]: A guidance from @ericstj as a preferred way of doing this

Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewColumn.cs]: https://github.com/dotnet/winforms/pull/1425#issuecomment-515136617

Comment (dreddy-work) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewColumn.cs]: We do generate 'TypeForwards' for all public types.  we were not creating typeforwards for internals/privates. Wanted to understand how this is different for  internal types used as member attributes.

File: src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/DataMemberFieldConverter.cs (hunk 2 above)
Comment (dreddy-work) [reply src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/DataMemberFieldConverter.cs]: fixing in next commit.

**Comment by zsd4yr** [comment]
Per conversation with @ericstj 	, @dreddy-work 	make sure you reference the previous assembly in the system.design facade project references.

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewColumn.cs (hunk 1 above)
Comment (dreddy-work) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewColumn.cs]: @ericstj , Can you please help us here understand on the scenarios that require explicit type forwards for internals?

**Comment by dreddy-work** [comment]
> Per conversation with @ericstj 	, @dreddy-work 	make sure you reference the previous assembly in the system.design facade project references.

It is already referenced and that is how typeforwards are generated for public interfaces/APIs.  @zsd4yr , i did not get your suggestion.

**Comment by ericstj** [comment]
I was just explaining how it works to @zsd4yr as far as the things to check to make sure any new typeforward would work.  If you're already referencing the assembly to-which the type was added its fine.

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewColumn.cs (hunk 1 above)
Comment (dreddy-work) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewColumn.cs]: @RussKie, adding type forwards manually may not be a desirable thing in the long run. especially for privates/internals ( we may remove them, for a reason, and go out of synch with type forwards that were added manually ). Unless there is a scenario/reason that the current (PR) way doesn't work, this seems cleaner approach.

Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewColumn.cs]: Sure thing 👍

Comment (dreddy-work) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewColumn.cs]: Okay. Please unblock your review if no further comments.

**Comment by dreddy-work** [comment]
@RussKie , can you please help merge this according to the core timelines? I am out of synch with bar check at this point.

**Pull request merged**